File: backend/server/pkg/cloude/cloude.go
package cloude

import (
	"app/internal/domain"
	"bytes"
	"encoding/json"
	"fmt"

	//"fmt"
	"net/http"
)

type FileData struct {
	Name   string `json:"name"`
	Format string `json:"format"`
	Path   string `json:"path"`
}
type DirData struct {
	Name string `json:"name"`
	Path string `json:"path"`
}

type WriteData struct {
	File FileData `json:"file"`
	Data string   `json:"data"`
}

type Cloude struct {
	Ipv4Address string
	Port        string
}

type IsExists struct {
	IsExists bool `json:"IsExists"`
}

type ResponeData struct {
	Data     string   `json:"data"`
	DataMass []string `json:"data_mass"`
}

type Location struct {
	Path string `json:"Path" `
}

func NewCloude(address string, port string) *Cloude {
	return &Cloude{Ipv4Address: address, Port: port}
}

func (c *Cloude) IsFile(data FileData) bool {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return false
	}
	client := http.Client{}
	req, err := http.NewRequest("GET", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/is-file", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")
	if err != nil {
		fmt.Println(err)
		return false
	}
	respone, _ := client.Do(req)
	var (
		exists IsExists
	)
	err = json.NewDecoder(respone.Body).Decode(&exists)
	if err != nil {
		fmt.Println(err)
		return false
	}

	respone.Body.Close()
	return exists.IsExists
}

func (c *Cloude) IsDir(data DirData) bool {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return false
	}
	client := http.Client{}
	req, err := http.NewRequest("GET", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/is-dir", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")

	respone, _ := client.Do(req)
	if err != nil {
		fmt.Println(err)
		return false
	}
	var (
		exists IsExists
	)
	err = json.NewDecoder(respone.Body).Decode(&exists)
	if err != nil {
		fmt.Println(err)
		return false
	}
	respone.Body.Close()
	return exists.IsExists
}

func (c *Cloude) CreateFile(data FileData) {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return
	}
	_, err = http.Post("http://"+c.Ipv4Address+":"+c.Port+"/cloude/create-file", "application/json", bytes.NewBuffer(byte_data))
	if err != nil {
		fmt.Println(err)
		return
	}
}

func (c *Cloude) CreateDir(data DirData) {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return
	}
	_, err = http.Post("http://"+c.Ipv4Address+":"+c.Port+"/cloude/create-dir", "application/json", bytes.NewBuffer(byte_data))
	if err != nil {
		fmt.Println(err)
		return
	}
}

func (c *Cloude) WriteFile(data WriteData) domain.Error {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	_, err = http.Post("http://"+c.Ipv4Address+":"+c.Port+"/cloude/write-file", "application/json", bytes.NewBuffer(byte_data))
	if err != nil {
		fmt.Println(err)
		return domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	return domain.Error{
		Error:  "Null",
		Objcet: nil,
	}
}

func (c *Cloude) ReadFile(data FileData) (string, domain.Error) {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return "", domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	client := http.Client{}
	req, err := http.NewRequest("GET", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/read-file", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")

	respone, _ := client.Do(req)
	if err != nil {
		fmt.Println(err)
		return "", domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	var (
		respone_data ResponeData
	)

	err = json.NewDecoder(respone.Body).Decode(&respone_data)
	if err != nil {
		fmt.Println(err)
		return "", domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	respone.Body.Close()
	return respone_data.Data, domain.Error{
		Error:  "Null",
		Objcet: nil,
	}
}

func (c *Cloude) DeleteFile(data FileData) domain.Error {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	client := http.Client{}
	req, err := http.NewRequest("DELETE", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/delete-file", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")
	respone, _ := client.Do(req)
	respone.Body.Close()

	return domain.Error{
		Error:  "Null",
		Objcet: nil,
	}
}

func (c *Cloude) DeleteDir(data FileData) domain.Error {
	byte_data, err := json.Marshal(data)
	if err != nil {
		fmt.Println(err)
		return domain.Error{
			Error:  err.Error(),
			Objcet: err,
		}
	}
	client := http.Client{}
	req, err := http.NewRequest("DELETE", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/delete-dir", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")
	respone, _ := client.Do(req)
	respone.Body.Close()
	return domain.Error{
		Error:  err.Error(),
		Objcet: nil,
	}
}

func (c *Cloude) DirNameAll(path Location) []string {
	byte_data, err := json.Marshal(path)
	if err != nil {
		fmt.Println(err)
		return []string{}
	}

	client := http.Client{}
	req, _ := http.NewRequest("GET", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/dir-name-all", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")
	respone, err := client.Do(req)
	if err != nil {
		fmt.Println(err)
		return []string{}
	}

	var (
		string_data ResponeData
	)

	err = json.NewDecoder(respone.Body).Decode(&string_data)
	if err != nil {
		fmt.Println(err)
		return []string{}
	}
	return string_data.DataMass
}

func (c *Cloude) FileNameAll(path Location) []string {
	byte_data, err := json.Marshal(path)
	if err != nil {
		fmt.Println(err)
		return []string{}
	}

	client := http.Client{}
	req, _ := http.NewRequest("GET", "http://"+c.Ipv4Address+":"+c.Port+"/cloude/file-name-all", bytes.NewBuffer(byte_data))
	req.Header.Set("Content-Type", "application/json")
	respone, err := client.Do(req)
	if err != nil {
		fmt.Println(err)
		return []string{}
	}

	var (
		string_data ResponeData
	)

	err = json.NewDecoder(respone.Body).Decode(&string_data)
	if err != nil {
		fmt.Println(err)
		return []string{}
	}
	return string_data.DataMass
}

File: backend/server/src/components/cloude/cloude.cpp
/*

 ________  ___       ________  ___  ___  ________  _______      
|\   ____\|\  \     |\   __  \|\  \|\  \|\   ___ \|\  ___ \     
\ \  \___|\ \  \    \ \  \|\  \ \  \\\  \ \  \_|\ \ \   __/|    
 \ \  \    \ \  \    \ \  \\\  \ \  \\\  \ \  \ \\ \ \  \_|/__  
  \ \  \____\ \  \____\ \  \\\  \ \  \\\  \ \  \_\\ \ \  \_|\ \ 
   \ \_______\ \_______\ \_______\ \_______\ \_______\ \_______\
    \|_______|\|_______|\|_______|\|_______|\|_______|\|_______|
                                                                
                                                            
*/


#if _WIN32 || __linux__
    #if _WIN32
        #include <WinSock2.h>
        #pragma comment(lib,"ws2_32.lib")

    #elif __linux__
        #include <sys/socket.h>
    #endif
#endif

#include "cloude.h"
#include "tools/json.hpp"

namespace Net{
    std::string get_data(std::string& data);
    std::string new_requset(std::string method,std::string address,std::string ip,int32_t port ,std::string type,std::string data);

    struct Respone {
        std::string respone;
        int32_t code;
        Respone(std::string data){
            nlohmann::json object = nlohmann::json::parse(data);
            this->respone = object["Data"];
            this->code = object["Code"];
        }
    };
}


bool Cloude::CreateFileCloude(FileData file){
    std::string dump = this->get_dump_file(file);
    Net::Respone data = Net::new_requset("POST","create-file",this->m_ip,this->m_port,"application/json",dump);
    return data.code == 200;
}

bool Cloude::CreateDir(DirData dir){
    std::string dump = this->get_dump_dir(dir);
    Net::Respone data = Net::new_requset("POST","create-dir",this->m_ip,this->m_port,"application/json",dump);
    return data.code == 200;
}

bool Cloude::DeleteFile(FileData file){
    std::string dump = this->get_dump_file(file);
    Net::Respone data = Net::new_requset("DELETE","delete-file",this->m_ip,this->m_port,"application/json",dump);
    return data.code == 200;
}
bool Cloude::DeleteDir(DirData dir){
    std::string dump = this->get_dump_dir(dir);
    Net::Respone data = Net::new_requset("DELETE","delete-dir",this->m_ip,this->m_port,"application/json",dump);
    return data.code  == 200;
}

bool Cloude::IsFile(FileData file) {
    std::string dump = this->get_dump_file(file);
    Net::Respone data = Net::new_requset("GET","is-file",this->m_ip,this->m_port,"application/json",dump);
    return strstr(data.respone.c_str(),"True");
}

bool Cloude::IsDir(DirData dir) {
    std::string dump = this->get_dump_dir(dir);
    Net::Respone data = Net::new_requset("GET","is-dir",this->m_ip,this->m_port,"application/json",dump);
    return strstr(data.respone.c_str(),"True");
}

bool Cloude::WriteFile(FileData file){
    std::string dump = this->get_dump_file(file);
    Net::Respone data = Net::new_requset("POST","write-file",this->m_ip,this->m_port,"application/json",dump);
    return data.code == 200; 
}

std::string Cloude::ReadFile(FileData file){
    std::string dump = this->get_dump_file(file);
    Net::Respone data = Net::new_requset("POST","read-file",this->m_ip,this->m_port,"application/json",dump);
    return data.respone; 
}

std::vector<std::string> Cloude::GetListFile(const char* path){
    nlohmann::json object =  {
        {"path",path}
    }; 
    Net::Respone data = Net::new_requset("GET","file-name-all",this->m_ip,this->m_port,"application/json",object.dump());

    std::vector<std::string> list;
    nlohmann::json respone = nlohmann::json::parse(data.respone);
    for(int i = 0; i < respone.size();i++)
        list.push_back(respone[i]);
    
    return list;
}

std::vector<std::string> Cloude::GetListDir(const char* path){
    nlohmann::json object =  {
        {"path",path}
    }; 
    Net::Respone data = Net::new_requset("GET","dir-name-all",this->m_ip,this->m_port,"application/json",object.dump());

    std::vector<std::string> list;
    nlohmann::json respone = nlohmann::json::parse(data.respone);
    for(int i = 0; i < respone.size();i++)
        list.push_back(respone[i]);
    
    return list;
}

 std::string Cloude::get_dump_file(FileData& file){
    nlohmann::json object = {
        {"path", file.file_path},
        {"type", file.file_type},
        {"name", file.file_name}
    };
    return object.dump();
 }

std::string Cloude::get_dump_dir(DirData& dir){
    nlohmann::json object = {
        {"path", dir.dir_path},
        {"name", dir.dir_name}
    };
    return object.dump();
 }

namespace Net {
    std::string get_data(std::string& respone){
        std::string buffer;
        bool isRead = false;
            
        for(int i = 0;i < respone.size(); i++){
            if(!isRead)
                if(respone[i] != '{' && respone[i] != '[')
                    continue;
                
            isRead = true;
            buffer+=respone[i];

        }
        return buffer;
    }
    std::string new_requset(std::string method,std::string address,std::string ip,int32_t port ,std::string type,std::string data){
        #if _WIN32
            WSADATA w_data;
            WORD word = MAKEWORD(2,2);
            WSAStartup(word,&w_data);
        #endif

        sockaddr_in sock_address;
        sock_address.sin_family = AF_INET;

        sock_address.sin_port   =  htons(port);
        sock_address.sin_addr.S_un.S_addr = inet_addr(ip.c_str());

            
        SOCKET client = socket(AF_INET,SOCK_STREAM,0);

        int is_connect = connect(client,(sockaddr*)&sock_address,sizeof(sock_address));
                
        std::string send_data;
              
                
                std::string header = method + " /" + address + " HTTP/1.0\n" +
                         "Host: " + ip + ":" + std::to_string(port) + "\n" +
                         "Content-Type: " + type + "; charset=utf-8\n" +
                         "Content-Length: " + std::to_string(data.size()) + "\r\n\r\n";

                send_data+= header + data;

                int is_send = send(client,send_data.c_str(),send_data.size(),0);
                char respone_buffer[65535];

                int size = recv(client,respone_buffer,sizeof(respone_buffer),0);
               
                #if _WIN32
                    WSACleanup();
                    closesocket(client);
                #elif __liunx__
                    close(clinet);
                #endif
                
                return Net::get_data(std::string(respone_buffer,size));        
        
    }
}